cpu/native/ng_tapnet/ng_tapnet.c

```diff
@@ -0,0 +1,615 @@
+    dev->event_cb = NULL;
+
+    return 0;
+}
+
+static int _get(ng_netdev_t *dev, ng_netconf_opt_t opt, void *value,
+                size_t *value_len)
+{
+    DEBUG("ng_tapnet: get ");
+
+    if ((dev == NULL) || (dev->driver == &ng_tapnet_driver)) {
+        DEBUG("[wrong device descriptor]\n");
+        return -ENODEV;
+    }
+
+    switch (opt) {
+        case NETCONF_OPT_ADDRESS:
+            DEBUG("address\n");
+            return _get_addr((ng_tapnet_t *)dev, value, value_len);
+
+        case NETCONF_OPT_ADDR_LEN:
+            DEBUG("address length\n");
+            return _get_addr_len(value, value_len);
+
+        case NETCONF_OPT_MAX_PACKET_SIZE:
+            DEBUG("maximum packet size\n");
+            return _get_max_pkt_sz(value, value_len);
+
+        case NETCONF_OPT_PROMISCUOUSMODE:
+            DEBUG("promiscous mode\n");
+            return _get_promiscousmode((ng_tapnet_t *)dev, value, value_len);
+
+        default:
+            DEBUG("[not supported: %d]\n", opt);
+            return -ENOTSUP;
+    }
+}
+
+static int _set(ng_netdev_t *dev, ng_netconf_opt_t opt, void *value,
+                size_t value_len)
+{
+    DEBUG("ng_tapnet: set ");
+
+    if ((dev == NULL) || (dev->driver == &ng_tapnet_driver)) {
+        DEBUG("[wrong device descriptor]\n");
+        return -ENODEV;
+    }
+
+    switch (opt) {
+        case NETCONF_OPT_PROMISCUOUSMODE:
+            DEBUG("promiscous mode\n");
+            return _set_promiscousmode((ng_tapnet_t *)dev, value, value_len);
+
+        default:
+            DEBUG("[not supported: %d]\n", opt);
+            return -ENOTSUP;
+    }
+}
+
+static void _rx_event(ng_tapnet_t *dev)
+{
+    uint8_t buffer[ETHER_MAX_LEN];
+    int nread = real_read(dev->tap_fd, buffer, ETHER_MAX_LEN);
+
+    DEBUG("ng_tapnet: read %d bytes\n", nread);
+
+    if (nread > 0) {
+        struct ether_header *hdr = (struct ether_header *)buffer;
+        ng_pktsnip_t *pkt;
+        size_t data_len = (nread - ETHER_HDR_LEN);
+        data_len += sizeof(ng_netif_hdr_t) + (2 * ETHER_ADDR_LEN);
+
+        if (!(dev->promiscous) && (memcmp(hdr->ether_dhost, dev->addr, ETHER_ADDR_LEN) != 0)) {
+            DEBUG("ng_tapnet: received for %02x:%02x:%02x:%02x:%02x:%02x\n"
+                  "That's not me => Dropped\n",
+                  hdr->ether_dhost[0], hdr->ether_dhost[1], hdr->ether_dhost[2],
+                  hdr->ether_dhost[3], hdr->ether_dhost[4], hdr->ether_dhost[5]);
+        }
+
+        pkt = ng_pktbuf_add(NULL, NULL, data_len, NG_NETTYPE_UNDEF);
+
+        if (pkt == NULL) {
+            DEBUG("ng_tapnet: no space left in packet buffer\n");
+            return;
+        }
+
+        ng_netif_hdr_init(pkt->data, ETHER_ADDR_LEN, ETHER_ADDR_LEN);
+        ng_netif_hdr_set_src_addr(pkt->data, hdr->ether_shost, ETHER_ADDR_LEN);
+        ng_netif_hdr_set_dst_addr(pkt->data, hdr->ether_dhost, ETHER_ADDR_LEN);
+        ((ng_netif_hdr_t *)pkt->data)->if_pid = thread_getpid();
+        /* XXX: Ethertype? */
```

Comment:
@haukepetersen How do we solve this? How does a L2 thread now where to send the data?